### DISCUSSION TITLE
Changing Key count mid-song?

### DISCUSSION THREAD

**Discussion by Onemario1234** [Leather128/LeathersFunkinEngine#28] [Ideas]
One addition I could see drawing people in is the ability to change the key count in the middle of a song, maybe even several times. For example, having a song that gradually expands the number of keys as the song goes on, or even having the key count change at random.

I've no skill in coding, so I don't know how easy or hard it would be to implement this, but I could see it being a cool feature. Would definitely catch my interest.

If you do add this kind of feature, it may be best to change the keybind display (the one that normally displays the keybinds at the start of songs with extra keys) to be out of the way of the arrows whenever the key count changes.

**Comment by Leather128** [comment]
I was thinking of adding this a while ago, but the problem i ran into is that, if you go from a higher key counter to a lower one, what do you do with the arrows that go past the amount of keys the game is at now?

**Reply by Onemario1234** [reply]
Maybe the notes are given random, new positions?

On the other side of the spectrum, perhaps when you're making the chart, you can choose the key count "per-section," so you can specifically write charts that change key mid-song as an intended part.

**Reply by Leather128** [reply]
that makes sense actually, although i was more thinking of it because i know the arrows will spawn a bit *before* each section, causing some arrows to be in a weird state (aka like going from 6 => 7 key and having a note from 7 key show up before it is 7 key)

**Reply by Onemario1234** [reply]
That's true. Perhaps the notes take on a different position as they scroll up before the keys change?

Or maybe they take on the exact position they'd be in after the key change before the keys even change.

For instance, if you're going from 4 to 5 key in a song with slow scroll speed, you'd see a diamond scrolling up for you to hit, with the key count changing to 5 to fit it just before it hits, depending on the placement. Not sure if I described that the best way I could've, but I hope you get what I mean. I'm basically saying that I think showing the upcoming notes as the new key layout would probably be the best solution.

**Reply by Onemario1234** [reply]
Semi-off topic, but perhaps it would be best to refer to the key count as "Mania," as that's often what I see others refer to it as. For example, changing key-count would be referred to as "changing the mania."

**Reply by Leather128** [reply]
personally for me, i like using key count because it's more user friendly, this is mainly because the mania value doesnt always line up with the amount of keys, so having it with a different name just made more sense to me lol.

**Reply by Onemario1234** [reply]
Fair enough.

Regardless, how's that solution?: Showing the upcoming notes as the new layout?

**Reply by Onemario1234** [reply]
I don't know how hard that would be to implement from a programming perspective, but it would work from a visual standpoint.

**Reply by Leather128** [reply]
That would work *visually* but programming wise, it'd be a lot harder because of how the x position of notes are coded.

**Reply by Onemario1234** [reply]
Yeah, that's a problem. Wish I could help, but I know zilch about modding.

**Comment by Onemario1234** [comment]
I can only imagine how much hell this would be to program, but now I'm just imagining a song that changes the key count for your opponent, but not you, or vice versa. Different key count on a per-player basis as well.

Not asking for this to be implemented. It just sounds insane.
